plugin/src/main/java/org/autorefactor/refactoring/rules/MapEliminateKeySetCallsRefactoring.java

```diff
@@ -50,39 +63,140 @@ public String getName() {
+                newFor.setExpression(b.invoke(b.copy(mapExpression), "entrySet"));
+                newFor.setParameter(b.declareSingleVariable(b.simpleName("entry"), "Map.Entry",
+                        getFriendlyTypeName(parameter.getType().resolveBinding()), getFriendlyTypeName(valueType)));
+
+                final Statement newBody = b.copy(enhancedFor.getBody());
+
+                // Declare 'KeyType' 'oldLoopVariable' = entry.getKey();
+                final VariableDeclarationStatement newKeyDeclaration = b.declareStmt(
+                        getFriendlyTypeName(parameter.getType().resolveBinding()),
+                        b.copy(parameter.getName()),
+                        b.invoke(b.name("entry"), "getKey"));
+
+                newFor.setBody(newBody);
+                ctx.getRefactorings().insertBefore(newKeyDeclaration,
+                        (ASTNode) ((Block) enhancedFor.getBody()).statements().get(0));
+                ctx.getRefactorings().replace(enhancedFor, newFor);
+                return DO_NOT_VISIT_SUBTREE;
+            }
+        }
+
+        return VISIT_SUBTREE;
+    }
+
+    /**
+     * @param typeBinding
+     * @return a short name, for classes in the java. packages, otherwise a fully qualified name.
+     */
+    private String getFriendlyTypeName(ITypeBinding typeBinding) {
+        final String name = typeBinding.getQualifiedName();
+        if (name.startsWith("java.")) {
+            return typeBinding.getName();
+        } else {
+            // TODO: fix to import the necessary type, if not found
+            return name;
+        }
+    }
+
+    /**
+     * Class to find map.get(loopVariable) construct in the AST tree, and collect the type of the value,
+     * which is unknown till one is located.
+     *
+     */
+    class FindExpression extends ASTVisitor {
+
+        private final Expression mapExpression;
+        private final SingleVariableDeclaration parameter;
+        private ITypeBinding valueType;
+
+        public FindExpression(Expression mapExpression, SingleVariableDeclaration parameter) {
+            this.mapExpression = mapExpression;
+            this.parameter = parameter;
+        }
+
+        @Override
+        public boolean visit(MethodInvocation node) {
+            if (ASTHelper.isSameVariable(node.getExpression(), mapExpression)) {
+                if (isMethod(node, "java.util.Map", "get", "java.util.Object")) {
+                    final ASTNode getArgument = (ASTNode) node.arguments().get(0);
+                    if (ASTHelper.isSameVariable(getArgument, parameter.getName())) {
+
+                        // collect the value type
+                        valueType = node.resolveTypeBinding();
+                        final ASTBuilder b = ctx.getASTBuilder();
+                        ctx.getRefactorings().replace(node, b.invoke(b.name("entry"), "getValue"));
+                        return DO_NOT_VISIT_SUBTREE;
```

Comment:
In addition, if `node.resolveTypeBinding()` (line 174), then we have already engaged in refactoring this code, but the rest of the refactoring will not be performed.
This will result in code that does not compile.